### PR TITLE
Metric docs

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Update to the metric docs:
* fixes some errors (functional interface does not automatically sync across ddp)
* add section on all the new metric hooks
* add section on how to get aggregated metric values

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/metrics.rst [hunk 1]
```diff
@@ -9,45 +9,55 @@
...
 =======
-This is a general package for PyTorch Metrics. These can also be used with regular non-lightning PyTorch code.
-Metrics are used to monitor model performance.
+Metrics are quantitative measures that can be used to monitor model performance,
+and they therefore are essential ingridient in any machine learning project. They
+can be used to measure performance during training, comparing different models and
+asses pitfalls such as overfitting.
 
-In this package, we provide two major pieces of functionality.
+The `pytorch lightning metric package` offers a general package for PyTorch Metrics.
```
Comment (justusschock) [inline docs/source/metrics.rst:17]:
```suggestion
The `pytorch_lightning.metrics package` offers a general package for PyTorch Metrics.
```

File: docs/source/metrics.rst [hunk 2]
```diff
@@ -9,45 +9,55 @@
...
     The metrics package is still in development! If we're missing a metric or you find a mistake, please send a PR!
-    to a few metrics. Please feel free to create an issue/PR if you have a proposed metric or have found a bug.
 
 ----------------
 
 Implement a metric
 ------------------
-You can implement metrics as either a PyTorch metric or a Numpy metric (It is recommended to use PyTorch metrics when possible,
-since Numpy metrics slow down training).
+While lightning provides a collection of standard used metrics, it is also possible
```
Comment (justusschock) [inline docs/source/metrics.rst:47]:
```suggestion
While pytorch lightning provides a collection of standard used metrics, it is also possible
```

File: docs/source/metrics.rst [hunk 3]
```diff
@@ -72,20 +86,131 @@ Here's an example showing how to implement a NumpyMetric
...
     :noindex:
 
-----------------
+
+Metric hooks
+^^^^^^^^^^^^
+
+Similar to a standard `torch.nn.Module`, the only *nessesary* method that should
+be implemented for a specific metric is the ``forward`` method. In this case, output
+we automatically be collected and averaged. That said, to gain fine control over
```
Comment (justusschock) [inline docs/source/metrics.rst:106]:
```suggestion
will automatically be collected and averaged. That said, to gain fine control over
```

File: docs/source/metrics.rst [hunk 4]
```diff
@@ -72,20 +86,131 @@ Here's an example showing how to implement a NumpyMetric
...
+""""""""""
+
+.. code-block::
+
+    @staticmethod
+    def ddp_reduce(self, data: Any, output: Any):
+
+Post-hook that implements how output from multiple devices should be collected and
+aggregated. We do not recommend overriding this, but instead consider overriding
+the two sub-methods called inside this hook: ``ddp_sync`` and ``aggregate``.
```
Comment (justusschock) [inline docs/source/metrics.rst:152]:
```suggestion
the two sub-methods called inside this hook: ``ddp_sync`` and ``aggregate``. If however, your operation for reduction can be done during syncing (e.g. via ``all_reduce``), this is the correct place to overwrite.
```

File: docs/source/metrics.rst [hunk 5]
```diff
@@ -72,20 +86,131 @@ Here's an example showing how to implement a NumpyMetric
...
+
+    @staticmethod
+    def compute(self, data: Any, output: Any):
+
+Post-hook that can be used to implement computations that needs to happen after
+output has been synced between devices.
+
+-------
+
+To summaries, in most cases it should be sufficient to implement ``forward`` (pre-ddp)
```
Comment (justusschock) [inline docs/source/metrics.rst:188]:
```suggestion
To summarize, in most cases it should be sufficient to implement ``forward`` (pre-ddp)
```

File: docs/source/metrics.rst [hunk 6]
```diff
@@ -122,6 +247,33 @@ These metrics even work when using distributed training:
     # any metric automatically reduces across GPUs (even the ones you implement using Lightning)
     trainer.fit(model)
 
+Class metrics aggregate both over multi device and multiple batches. The aggregated
+value can be access through the `metric.aggregated` property. When this property is
```
Comment (justusschock) [inline docs/source/metrics.rst:251]:
```suggestion
Class metrics aggregate both over multi device and multiple batches. The aggregated
value over batches can be access through the `metric.aggregated` property. The aggregated value over multiple devices in the same state is returned upon module call. When this property is
```

**Review by justusschock — APPROVED** [review]
